### PR TITLE
Multiple dockercfg

### PR DESCRIPTION
**Pull request by enoodle**
Adds support for multiple dockercfg files defined by the user with repeated flags: `--dockercfg file1 --dockercfg file2`. This will try to pull the image with all the authentications from all the files. (Assuming the user is not particular with which authentication he wants to use otherwise he would have only sent that one..):

`sudo ./image-inspector --image=registry.access.redhat.com/openshift3/metrics-cassandra --chroot --serve=0.0.0.0:5000 --path=/tmp/image-content/ --dockercfg dockercfg1 --dockercfg dockercfg2`

### CODE REVIEW

**Comment by enoodle** [comment]
This is rebased on the split patch from @pweil- , I will rebase on master once it its merged. @pweil- I will appreciate a review, especially on the test file.

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -93,7 +93,7 @@ func (i *defaultImageInspector) Inspect() error {
 	container, err := client.CreateContainer(docker.CreateContainerOptions{
 		Name: randomName,
 		Config: &docker.Config{
-			Image:      i.opts.Image,
+			Image: i.opts.Image,
```
Comment (simon3z) [inline pkg/inspector/image-inspector.go]: @enoodle why this change?

File: cmd/image-inspector.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ func main() {
 	flag.StringVar(&inspectorOptions.DstPath, "path", inspectorOptions.DstPath, "Destination path for the image files")
 	flag.StringVar(&inspectorOptions.Serve, "serve", inspectorOptions.Serve, "Host and port where to serve the image with webdav")
 	flag.BoolVar(&inspectorOptions.Chroot, "chroot", inspectorOptions.Chroot, "Change root when serving the image with webdav")
-	flag.StringVar(&inspectorOptions.DockerCfg, "dockercfg", inspectorOptions.DockerCfg, "Location of the docker configuration file")
+	flag.StringVar(&inspectorOptions.DockerCfg, "dockercfg", inspectorOptions.DockerCfg, "Location of the docker configuration files seperated by ',' if more than one")
```
Comment (simon3z) [inline cmd/image-inspector.go]: @enoodle what about files with a comma `,` in their name? :smile: 

I think there should be a way to specify more than once the same flag, e.g.:

`image-inspector ... --dockercfg=config1 --dockercfg=config2 --dockercfg=config3 ...`

cc @pweil-

File: pkg/inspector/image-inspector.go (hunk 1 above)
Comment (enoodle) [inline pkg/inspector/image-inspector.go]: I am sorry, Its a problem with my vim and its lint.. its autocorrects this. I will look into this.. Oo

File: cmd/image-inspector.go (hunk 2 above)
Comment (pweil-) [inline cmd/image-inspector.go]: spf13/pflags makes it easy with `StringSliceVar` https://github.com/spf13/pflag/blob/master/string_slice.go#L67.  If we don't want to pull in that dependency it would have to be a custom `Var` with a set implementation.

File: pkg/inspector/image-inspector.go [hunk 3]
```diff
@@ -272,23 +272,37 @@ func generateRandomName() (string, error) {
...
-		reader, err := os.Open(dockercfg)
-		if err != nil {
-			return nil, fmt.Errorf("Unable to open docker config file: %v\n", err)
-		}
-		if imagePullAuths, err = docker.NewAuthConfigurations(reader); err != nil {
-			return nil, fmt.Errorf("Unable to parse docker config file: %v\n", err)
-		}
-		if len(imagePullAuths.Configs) == 0 {
-			return nil, fmt.Errorf("No auths were found in the given dockercfg file\n")
+		for _, dcfgFile := range strings.Split(dockercfg, ",") {
```
Comment (pweil-) [inline pkg/inspector/image-inspector.go]: If we do end up with a custom impl (and not pflags) let's make sure the flag parsing stays in the cmd/image-inspector file (or package if it is a util).  I've been (mostly) trying to follow the Complete-Validate-Run scheme of Kube: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-implementation-conventions

File: pkg/inspector/image-inspector_test.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+func TestGetAuthConfigs(t *testing.T) {
+
+	tests := map[string]struct {
+		dockercfg      string
+		username       string
+		passwordFile   string
+		len            int
+		shouldValidate bool
```
Comment (pweil-) [inline pkg/inspector/image-inspector_test.go]: there are no use cases that test the `false` value of this

Comment (enoodle) [inline pkg/inspector/image-inspector_test.go]: Thanks, you are right. I will change it to shouldFail and add a test for failure.

File: cmd/image-inspector.go (hunk 2 above)
Comment (enoodle) [inline cmd/image-inspector.go]: @simon3z I am leaning towards using pflags, how do you feel about adding dependencies? I don't think implementing a custom `Var` is too hard but I would like to avoid it if there is already a library :grin:

Comment (simon3z) [inline cmd/image-inspector.go]: @enoodle @pweil- let's follow what kubernetes/openshift is doing for its cmds. I remember there are several that take an array of strings.

**Comment by enoodle** [comment]
I rebased it on #3 just for convenience with future merge. Will rebase on master once #3 is merged.

File: pkg/inspector/image-inspector_test.go [hunk 5]
```diff
@@ -61,7 +62,46 @@ func TestScanImage(t *testing.T) {
 				}
 			}
 		}
+	}
+}
+
+func TestGetAuthConfigs(t *testing.T) {
+
```
Comment (simon3z) [inline pkg/inspector/image-inspector_test.go]: @enoodle extra while line

**Comment by simon3z** [comment]
@pweil- @enoodle this is next. Please re-test / re-review and let's get this in as well.

**Comment by enoodle** [comment]
=== RUN   TestValidate
--- PASS: TestValidate (0.00s)
PASS
coverage: 100.0% of statements
ok      github.com/simon3z/image-inspector/pkg/cmd      0.011s  coverage: 100.0% of statements
=== RUN   TestScanImage
2016/04/21 16:20:43 MockScanner scanning here. Placing results in 
2016/04/21 16:20:43 MockScanner scanning here. Placing results in 
2016/04/21 16:20:43 MockScanner scanning here. Placing results in 
--- PASS: TestScanImage (0.00s)
=== RUN   TestGetAuthConfigs
--- PASS: TestGetAuthConfigs (0.00s)
=== RUN   TestCreateOutputDir
--- PASS: TestCreateOutputDir (0.00s)
PASS
coverage: 22.5% of statements
ok      github.com/simon3z/image-inspector/pkg/inspector        0.005s  coverage: 22.5% of statements
=== RUN   TestGetRhelDist
--- PASS: TestGetRhelDist (0.00s)
=== RUN   TestScan
--- PASS: TestScan (0.00s)
PASS
coverage: 38.3% of statements
ok      github.com/simon3z/image-inspector/pkg/openscap 0.025s  coverage: 38.3% of statements

File: pkg/inspector/image-inspector_test.go [hunk 6]
```diff
@@ -61,7 +62,45 @@ func TestScanImage(t *testing.T) {
...
+	goodUserAndPass := iicmd.NewDefaultImageInspectorOptions()
+	goodUserAndPass.Username = "erez"
+	goodUserAndPass.PasswordFile = "test/passwordFile1"
+
+	tests := map[string]struct {
+		opts       *iicmd.ImageInspectorOptions
+		shouldFail bool
+	}{
+		"two dockercfg":             {opts: goodTwoDockerCfg, shouldFail: false},
+		"username and passwordFile": {opts: goodUserAndPass, shouldFail: false},
```
Comment (pweil-) [inline pkg/inspector/image-inspector_test.go]: should we add failing tests too?

File: pkg/cmd/types.go [hunk 7]
```diff
@@ -9,6 +9,19 @@ var (
 	ScanOptions = []string{"openscap"}
 )
 
+type MultiStringVar struct {
+	Values []string
+}
+
+func (sv *MultiStringVar) Set(s string) error {
```
Comment (pweil-) [inline pkg/cmd/types.go:17]: nit: This is more of an `add` than a `set`.  Have calls most of the morning but I'll take another look this afternoon.  If it's working and there are no other comments then don't wait on me. :+1:

**Comment by simon3z** [comment]
@enoodle for the record can you drop here in the PR comments how the command line looks like? Thanks.

File: pkg/cmd/types.go (hunk 7 above)
Comment (enoodle) [inline pkg/cmd/types.go:17]: In our case Yes. I am implementing an interface here for the `flag` library: https://golang.org/src/flag/flag.go?s=6274:6334#L230

Comment (pweil-) [inline pkg/cmd/types.go:17]: perfect, that is what I figured since I didn't see an explicit set anywhere in the other code

File: pkg/inspector/image-inspector_test.go (hunk 6 above)
Comment (enoodle) [inline pkg/inspector/image-inspector_test.go]: added

File: pkg/cmd/types.go (hunk 7 above)
Comment (enoodle) [inline pkg/cmd/types.go:17]: I added a comment to clarify that.

**Comment by enoodle** [comment]
@simon3z added example command line in the PR comment

**Comment by simon3z** [comment]
@pweil- LGTY?

**Comment by pweil-** [comment]
All my comments have been addressed.  LGTM.  Thanks @enoodle!

**Comment by enoodle** [comment]
@simon3z ping
